### PR TITLE
mobile: fixing explicit flow control bugs.

### PR DESCRIPTION
**Pull request by alyssawilk**
When deferring the error for the explicit flow control case, we didn't close the stream, which could result in attempting to send data to the weak pointer (now an ENVOY_BUG)

When deferring an error, we also never actually sent it.  Now forwarding it on unless the response is complete.

Risk Level: n/a (mobile only)
Testing: new tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: mobile/test/common/integration/client_integration_test.cc
Comment (abeyad) [inline mobile/test/common/integration/client_integration_test.cc:909]: I'm a bit confused by this: once the stream is reset (in L904), I thought we shouldn't be allowed to send data on the stream?  Or is this the client->server stream whereas L904 cancels server->client?

**Review by abeyad — APPROVED** [review]

File: mobile/test/common/integration/client_integration_test.cc
Comment (alyssawilk) [reply mobile/test/common/integration/client_integration_test.cc:909]: correct - once the stream is reset we shouldn't be allowed to send data on the stream.  
What can happen is
sendData is called but not executed
stream is reset, and reset is communicated to API
sendData is executed

from the mobile stack's point of view there's no way of differentiating between that and
stream is reset and reset is communicated to API
sendData is called
sendData is executed

the first set of ordering is almost impossible to guarantee so here we're intentionally breaking the API contract in order to test the race condition deterministcally.

**Comment by alyssawilk** [comment]
@RyanTheOptimist this has explicit mobile LGTM so I just need envoy maintainer LGTM for the minor Envoy test framework change (though feel free to review the whole thing iff you have time)

File: mobile/test/common/integration/client_integration_test.cc
Comment (abeyad) [reply mobile/test/common/integration/client_integration_test.cc:909]: got it, thanks for the explanation!

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**
